### PR TITLE
feat(config): Derive Shelley genesis from dingo_shelley_start_time /dingo_epoch_length_slots

### LINKED ISSUE #448: dingo: derive Shelley genesis from dingo_shelley_start_time / dingo_epoch_length_slots
## Problem

`internal/config/config.go` derives Byron and Shelley genesis parameters (start time, epoch length, slot length, K) from a hardcoded switch over named networks (`mainnet`/`preprod`/`preview`/`sancho`). On a Dingo devnet, custom testnet, or any future network, these defaults are wrong. The downstream effect is that `utils.go::getSlotTipRef` returns an incorrect tip slot, which corrupts the epoch progress bar and the tip-difference display.

Dingo (verified against v0.41.0) publishes the relevant genesis information on its `/metrics` endpoint:

- `dingo_shelley_start_time` — Shelley-era start time (unix seconds)
- `dingo_epoch_length_slots` — slots per Shelley epoch

For non-mainnet/preprod networks where the operator gave nview no hint, this is more authoritative than nview's hardcoded table.

## Solution

After the first successful Dingo scrape, override the genesis-derived config fields **only if the user has not explicitly set them** and the Dingo-published values are non-zero.

### New PromMetrics fields

```go
DingoShelleyStartTime  uint64 `json:"dingo_shelley_start_time"`
DingoEpochLengthSlots  uint64 `json:"dingo_epoch_length_slots"`
```

These should be added in the same change as the Dingo Diagnostics view (separate issue) to avoid two passes over `PromMetrics`.

### Override function

In `internal/config/config.go`:

```go
// ApplyDingoGenesisOverride mutates the singleton config when Dingo metrics
// supersede the hardcoded network table. Only fires when:
//   - both metric values are non-zero
//   - the corresponding config fields still match the network-table defaults
//     (i.e. the user did not pin them via env or YAML)
//
// Returns true if any field was changed.
func ApplyDingoGenesisOverride(shelleyStart, epochLenSlots uint64) bool
```

Detecting "user pinned the value" is tricky because envconfig writes into the same struct. Approach:

1. After `LoadConfig`, snapshot the post-load network defaults that would have been applied if no env vars were present (call `getDefaultConfig()` then re-run the four `populate*` helpers on it).
2. Store this snapshot as `defaultsForCurrentNetwork` on a package-level var.
3. In `ApplyDingoGenesisOverride`, only mutate a field if its current value equals the snapshot value for that field. If the user set `BYRON_GENESIS_START_SEC` or `SHELLEY_EPOCH_LENGTH`, the field will differ from the snapshot and the override skips.

### Mapping

- `dingo_shelley_start_time` → `cfg.Node.ByronGenesis.StartTime`. Devnets typically have no Byron era; treating Shelley start as Byron start with `ShelleyTransEpoch = 0` makes `getSlotTipRef` compute purely Shelley-era slots, which is correct for these networks. Document this behavior.
- `dingo_epoch_length_slots` → `cfg.Node.ShelleyGenesis.EpochLength`.
- `cfg.Node.ShelleyTransEpoch` → forced to `0` when override fires (Shelley-from-genesis assumption).

If a future Dingo also publishes a true Byron start time, swap these mappings. The override function stays the only place that needs to change.

### Invocation

In the main scrape loop in `main.go`, after `getPromMetrics` succeeds and `detectNodeType` has run:

```go
if getEffectiveNodeBinary() == DINGO_BINARY && metrics != nil {
    if config.ApplyDingoGenesisOverride(
        metrics.DingoShelleyStartTime,
        metrics.DingoEpochLengthSlots,
    ) {
        slog.Info("genesis params overridden from Dingo metrics",
            "shelleyStart", metrics.DingoShelleyStartTime,
            "epochLengthSlots", metrics.DingoEpochLengthSlots,
        )
    }
}
```

The log line is one-shot from the operator's perspective: subsequent calls are no-ops because the snapshot comparison stops matching after the first override.

## Acceptance criteria

- [ ] `ApplyDingoGenesisOverride` exists in `internal/config` and is package-tested
- [ ] Override fires when defaults are present and metric values are non-zero
- [ ] Override is a no-op when the user set any of: `BYRON_GENESIS_START_SEC`, `BYRON_EPOCH_LENGTH`, `BYRON_SLOT_LENGTH`, `SHELLEY_EPOCH_LENGTH`, `SHELLEY_SLOT_LENGTH`
- [ ] Override is a no-op when either metric is zero
- [ ] One-shot info-level log line emitted at first successful override
- [ ] Tip slot computed by `getSlotTipRef` after override matches Dingo-reported tip within ±1 slot on a fixture devnet

## Risk and rollback

Medium risk: silent slot-math regression if Dingo renames or removes the metrics. Mitigation:

- Override is gated on both fields being non-zero — removed metrics simply mean no override, and nview falls back to the hardcoded table (degraded but not broken)
- Log line surfaces the fact that an override happened, so a regression on a fresh Dingo build is visible in the in-TUI log pane

Rollback: revert the call site in the scrape loop. The function can stay dormant.

## Out of scope

- Reading Dingo's Blockfrost endpoint for genesis info (different transport, not warranted yet)
- Override for Amaru — Amaru does not currently publish equivalent metrics

### PR DESCRIPTION
**Pull request by arepala-uml**
1. Added parsing for Dingo genesis metrics dingo_shelley_start_time and dingo_epoch_length_slots.
2. Added ApplyDingoGenesisOverride in config to update slot config from dingo metrics where we preserved user-pinned Byron/Shelley genesis values during config population. Mapped dingo shelley start time to ByronGenesis.StartTime and dingo epoch length to ShelleyGenesis.EpochLength.
3. Added a current-network defaults snapshot to detect user-pinned genesis config.
4. Called the override after successful Dingo Prometheus scrapes and added a log statement once when dingo genesis params override hardcoded defaults.
5. Added tests for successful override, zero metrics, pinned config, and metric parsing.

Closes #448 

---
## Summary by cubic
Derives Shelley genesis from Dingo Prometheus metrics and applies overrides only when config still matches per-network defaults. Makes the override thread-safe and fixes slot math on Dingo devnets without a Byron era. Implements #448.

- New Features
  - Parse `dingo_shelley_start_time` and `dingo_epoch_length_slots` in `PromMetrics`.
  - Add `ApplyDingoGenesisOverride(...)`: when both metrics exist and values still match network defaults, set `ByronGenesis.StartTime`, `ShelleyGenesis.EpochLength`, and `ShelleyTransEpoch=0`; run after a `dingo` scrape and log on change.
  - Snapshot per-network defaults at load to detect pinned values and avoid overriding user config.

- Refactors
  - Use `atomic.Pointer` for the global config and defaults snapshot to prevent races during overrides.
  - Make `populateByronGenesis`/`populateShelleyGenesis` fill only missing fields and respect pinned values (including `K`).
  - Add tests for override success, zero metrics, pinned config (incl. pinned `K`), and metric parsing.

<sup>Written for commit cbd4cb522da22354cba2f5fb461865f9ce412493. Summary will update on new commits.</sup>

## Summary by CodeRabbit

* **New Features**
  * Automatic override of genesis timing from Dingo node metrics when available.

* **Improvements**
  * Configuration now maintains network-specific defaults to better distinguish user changes from computed defaults.
  * Genesis updates are more selective, preserving pinned/user-provided values while permitting metric-driven adjustments.

* **Tests**
  * Added unit tests covering Dingo metrics parsing and genesis-override behaviors.

[Review Change Stack](https://app.coderabbit.ai/change-stack/blinklabs-io/nview/pull/452)

### CODE REVIEW

**Comment by arepala-uml** [comment]
Testing part:

I have tested the successful override, zero metrics, pinned config with the three added unit-tests.
Also, I have validated manually with my local dingo and added a small test to validate whether genesis metrics are overwriting or not

Firstly I have checked whether the metrics are there or not with local dingo prometheus where i have just did a curl request to see the actual values of dingo_shelley_start_time, dingo_epoch_length_slots

```
akhil-mac@Akhils-MacBook-Air nview %  curl -s http://127.0.0.1:12798/metrics | grep dingo_shelley
# HELP dingo_shelley_start_time Shelley genesis start as unix timestamp
# TYPE dingo_shelley_start_time gauge
dingo_shelley_start_time{network="preview"} 1.666656e+09
akhil-mac@Akhils-MacBook-Air nview % 
akhil-mac@Akhils-MacBook-Air nview % 
akhil-mac@Akhils-MacBook-Air nview %  curl -s http://127.0.0.1:12798/metrics | grep dingo_epoch  
# HELP dingo_epoch_length_slots slots per epoch for the current network
# TYPE dingo_epoch_length_slots gauge
dingo_epoch_length_slots{network="preview"} 86400
akhil-mac@Akhils-MacBook-Air nview % 
akhil-mac@Akhils-MacBook-Air nview % 
```

After that, I have written a small test by pointing my dingo binary with local host & local prometheus port. After that, I have collected the prometheus metrics and next I logged the above two metric values. After that, I called the newly added ApplyDingoGenesisOverride function to see whether config gets overriding or not.

```
package main

import (
	"context"
	"testing"

	"github.com/blinklabs-io/nview/internal/config"
)


func TestManualDingoGenesisOverrideLocal(t *testing.T) {
	cfg, err := config.LoadConfig("")
	if err != nil {
		t.Fatalf("LoadConfig() error = %v", err)
	}

	cfg.Node.Binary = DINGO_BINARY
	cfg.Prometheus.Host = "127.0.0.1"
	cfg.Prometheus.Port = 12798

	metrics, err := getPromMetrics(context.Background())
	if err != nil {
		t.Fatalf("getPromMetrics() error = %v", err)
	}

	t.Logf(
		"dingo metrics: shelleyStart=%d epochLengthSlots=%d slotNum=%d",
		metrics.DingoShelleyStartTime,
		metrics.DingoEpochLengthSlots,
		metrics.SlotNum,
	)

	if !config.ApplyDingoGenesisOverride(
		metrics.DingoShelleyStartTime,
		metrics.DingoEpochLengthSlots,
	) {
		t.Fatal("ApplyDingoGenesisOverride() = false, expected true")
	}

	updated := config.GetConfig()

	t.Logf(
		"updated config: byronStart=%d shelleyEpochLength=%d shelleyTransEpoch=%d",
		updated.Node.ByronGenesis.StartTime,
		updated.Node.ShelleyGenesis.EpochLength,
		updated.Node.ShelleyTransEpoch,
	)

	if updated.Node.ByronGenesis.StartTime != metrics.DingoShelleyStartTime {
		t.Fatal("ByronGenesis.StartTime was not updated from Dingo metric")
	}
	if updated.Node.ShelleyGenesis.EpochLength != metrics.DingoEpochLengthSlots {
		t.Fatal("ShelleyGenesis.EpochLength was not updated from Dingo metric")
	}
	if updated.Node.ShelleyTransEpoch != 0 {
		t.Fatal("ShelleyTransEpoch was not forced to 0")
	}
}
```

Test result

```
akhil-mac@Akhils-MacBook-Air nview % go test -run TestManualDingoGenesisOverrideLocal -count=1 -v

=== RUN   TestManualDingoGenesisOverrideLocal
    dingo_manual_test.go:39: dingo metrics: shelleyStart=1666656000 epochLengthSlots=86400 slotNum=1362603
    dingo_manual_test.go:55: updated config: byronStart=1666656000 shelleyEpochLength=86400 shelleyTransEpoch=0
--- PASS: TestManualDingoGenesisOverrideLocal (0.01s)
PASS
ok      github.com/blinklabs-io/nview   0.261s
akhil-mac@Akhils-MacBook-Air nview % go test -run TestManualDingoGenesisOverrideLocal -count=1 -v

=== RUN   TestManualDingoGenesisOverrideLocal
    dingo_manual_test.go:39: dingo metrics: shelleyStart=1666656000 epochLengthSlots=86400 slotNum=1583653
    dingo_manual_test.go:55: updated config: byronStart=1666656000 shelleyEpochLength=86400 shelleyTransEpoch=0
--- PASS: TestManualDingoGenesisOverrideLocal (0.01s)
PASS
ok      github.com/blinklabs-io/nview   0.259s
akhil-mac@Akhils-MacBook-Air nview % 
```

File: internal/config/config.go [hunk 1]
```diff
@@ -222,73 +294,85 @@ func (c *Config) populateShelleyTransEpoch() error {
...
+		if c.Node.ByronGenesis.StartTime == 0 {
 			c.Node.ByronGenesis.StartTime = 1506203091
 		}
-	} else {
-		return errors.New("unable to populate byron genesis config")
 	}
-	c.Node.ByronGenesis.EpochLength = (10 * c.Node.ByronGenesis.K)
+	if c.Node.ByronGenesis.EpochLength == 0 {
+		c.Node.ByronGenesis.EpochLength = (10 * c.Node.ByronGenesis.K)
+		return nil
```
[the remark this answers is not in the document]
Comment (arepala-uml) [reply internal/config/config.go:354]: Addressed it in latest commit and please review it.

File: internal/config/config.go [hunk 2]
```diff
@@ -155,6 +163,70 @@ func GetConfig() *Config {
...
+		return false
+	}
+
+	changed := cfg.Node.ByronGenesis.StartTime != shelleyStart ||
+		cfg.Node.ShelleyGenesis.EpochLength != epochLenSlots ||
+		cfg.Node.ShelleyTransEpoch != 0
+
+	cfg.Node.ByronGenesis.StartTime = shelleyStart
+	cfg.Node.ShelleyGenesis.EpochLength = epochLenSlots
+	cfg.Node.ShelleyTransEpoch = 0
```
[the remark this answers is not in the document]
Comment (arepala-uml) [reply internal/config/config.go]: Addressed it in latest commit and please review it.

File: internal/config/config.go [hunk 3]
```diff
@@ -103,7 +103,10 @@ func getDefaultConfig() *Config {
 }
 
 // Singleton config instance with default values
-var globalConfig = getDefaultConfig()
+var (
+	globalConfig              = getDefaultConfig()
+	defaultsForCurrentNetwork = getDefaultConfig()
```
[the remark this answers is not in the document]
Comment (arepala-uml) [reply internal/config/config.go]: Addressed it in latest commit and please review it.

File: internal/config/config.go [hunk 4]
```diff
@@ -155,6 +163,70 @@ func GetConfig() *Config {
...
+	}
+
+	cfg := globalConfig
+	defaults := defaultsForCurrentNetwork
+	// Only override values that still match the network-table defaults.
+	if cfg.Node.ByronGenesis.StartTime != defaults.Node.ByronGenesis.StartTime ||
+		cfg.Node.ByronGenesis.EpochLength != defaults.Node.ByronGenesis.EpochLength ||
+		cfg.Node.ByronGenesis.SlotLength != defaults.Node.ByronGenesis.SlotLength ||
+		cfg.Node.ShelleyGenesis.EpochLength != defaults.Node.ShelleyGenesis.EpochLength ||
+		cfg.Node.ShelleyGenesis.SlotLength != defaults.Node.ShelleyGenesis.SlotLength {
```
[the remark this answers is not in the document]
Comment (arepala-uml) [reply internal/config/config.go]: Addressed it in latest commit and please review it.

File: internal/config/config.go [hunk 5]
```diff
@@ -155,6 +163,70 @@ func GetConfig() *Config {
...
+	}
+
+	changed := cfg.Node.ByronGenesis.StartTime != shelleyStart ||
+		cfg.Node.ShelleyGenesis.EpochLength != epochLenSlots ||
+		cfg.Node.ShelleyTransEpoch != 0
+
+	cfg.Node.ByronGenesis.StartTime = shelleyStart
+	cfg.Node.ShelleyGenesis.EpochLength = epochLenSlots
+	cfg.Node.ShelleyTransEpoch = 0
+	return changed
```
[the remark this answers is not in the document]
Comment (arepala-uml) [reply internal/config/config.go]: Made changes regarding the globalConfig to overcome race conditions and thanks for letting me know. Please review it.

File: main.go [hunk 6]
```diff
@@ -278,6 +278,20 @@ func main() {
 				}
 				continue
 			}
+			if getEffectiveNodeBinary() == DINGO_BINARY && prom != nil {
+				if config.ApplyDingoGenesisOverride(
```
[the remark this answers is not in the document]
Comment (arepala-uml) [reply main.go:282]: Addressed the race conditions and please review it.

**Review by wolf31o2 — APPROVED** [review]
